pkg/rtc/mediatrack.go

```diff
@@ -216,163 +197,27 @@ func (t *MediaTrack) AddSubscriber(sub types.Participant) error {
-	// when outtrack is bound, start loop to send reports
-	downTrack.OnBind(func() {
-		go subTrack.Bound()
-		go t.sendDownTrackBindingReports(sub)
-	})
-	downTrack.OnPacketSent(func(_ *sfu.DownTrack, size int) {
-		t.params.Telemetry.OnDownstreamPacket(subscriberID, t.ID(), size)
-	})
-	downTrack.OnPaddingSent(func(_ *sfu.DownTrack, size int) {
-		t.params.Telemetry.OnDownstreamPacket(subscriberID, t.ID(), size)
-	})
-	downTrack.OnRTCP(func(pkts []rtcp.Packet) {
-		t.params.Telemetry.HandleRTCP(livekit.StreamType_DOWNSTREAM, subscriberID, t.ID(), pkts)
-	})
-
-	downTrack.OnCloseHandler(func() {
-		go func() {
-			t.subscribedTracks.Delete(subscriberID)
-			t.params.Telemetry.TrackUnsubscribed(context.Background(), subscriberID, t.ToProto())
-
-			// ignore if the subscribing sub is not connected
-			if sub.SubscriberPC().ConnectionState() == webrtc.PeerConnectionStateClosed {
-				return
-			}
-
-			// if the source has been terminated, we'll need to terminate all of the subscribedtracks
-			// however, if the dest sub has disconnected, then we can skip
-			if sender == nil {
-				return
-			}
-			t.params.Logger.Debugw("removing peerconnection track",
-				"track", t.ID(),
-				"subscriber", sub.Identity(),
-				"subscriberID", subscriberID,
-				"kind", t.Kind(),
-			)
-			if err := sub.SubscriberPC().RemoveTrack(sender); err != nil {
-				if err == webrtc.ErrConnectionClosed {
-					// sub closing, can skip removing subscribedtracks
-					return
-				}
-				if _, ok := err.(*rtcerr.InvalidStateError); !ok {
-					// most of these are safe to ignore, since the track state might have already
-					// been set to Inactive
-					t.params.Logger.Debugw("could not remove remoteTrack from forwarder",
-						"error", err,
-						"subscriber", sub.Identity(),
-						"subscriberID", subscriberID,
-					)
-				}
-			}
-
-			t.NotifySubscriberMute(subscriberID)
-			sub.RemoveSubscribedTrack(subTrack)
-			sub.Negotiate()
-		}()
-	})
-	if t.Kind() == livekit.TrackType_AUDIO {
-		downTrack.AddReceiverReportListener(t.handleMaxLossFeedback)
-	}
-
-	t.subscribedTracks.Store(subscriberID, subTrack)
-	subTrack.SetPublisherMuted(t.IsMuted())
-
-	t.receiver.AddDownTrack(downTrack)
-	// since sub will lock, run it in a goroutine to avoid deadlocks
-	go func() {
-		t.NotifySubscriberMaxQuality(subscriberID, livekit.VideoQuality_HIGH) // start with HIGH, let subscription change it later
-		sub.AddSubscribedTrack(subTrack)
-		sub.Negotiate()
-	}()
-
-	t.params.Telemetry.TrackSubscribed(context.Background(), subscriberID, t.ToProto())
 	return nil
 }
 
 func (t *MediaTrack) NumUpTracks() (uint32, uint32) {
 	numExpected := atomic.LoadUint32(&t.numUpTracks)
 
-	t.maxQualityLock.RLock()
-	maxSubscribed := uint32(0)
-	if !t.allSubscribersMuted {
-		maxSubscribed = uint32(SpatialLayerForQuality(t.maxSubscribedQuality) + 1)
-	}
-	t.maxQualityLock.RUnlock()
-	if maxSubscribed < numExpected {
-		numExpected = maxSubscribed
+	numSubscribed := t.numSubscribed()
```

Comment:
nit: should this be renamed to `numSubscribedLayers`? It took me a minute to realize it's referencing # of layers instead of # of subscribers

Reply:
Will change name